### PR TITLE
Make various typechecking improvements

### PR DESCRIPTION
**Pull request by AndrewFerr**
Some type-checking improvements that make parts of the code more expressive & easier to work with.

Note that this is based on the config widget PR because it is the source of one of the type annotations added by this PR.

### CODE REVIEW

File: src/Widgets/BridgeWidgetApi.ts [hunk 1]
```diff
@@ -156,7 +156,7 @@ export class BridgeWidgetApi {
         if (!connection.provisionerUpdateConfig || !connection.getProvisionerDetails)  {
             throw new ApiError("Connection type does not support updates", ErrCode.UnsupportedOperation);
         }
-        await connection.provisionerUpdateConfig(req.userId, req.body);
+        connection.provisionerUpdateConfig(req.userId, req.body);
```
Comment (Half-Shot) [inline src/Widgets/BridgeWidgetApi.ts:159]: Did this stop being promisey?

**Review by Half-Shot — APPROVED** [review]
One comment, otherwise LGTM

File: src/Widgets/BridgeWidgetApi.ts (hunk 1 above)
Comment (AndrewFerr) [reply src/Widgets/BridgeWidgetApi.ts:159]: Yes: https://github.com/matrix-org/matrix-hookshot/blob/main/src/Connections/IConnection.ts#L68

**Comment by AndrewFerr** [comment]
Before submitting this, there's a type error added by this change that `yarn lint` missed which I'll try fixing.

https://github.com/matrix-org/matrix-hookshot/pull/505/files#diff-e9a8de0579bab9c5fea95a16abc38388d4f7b0712ea52c581238f6d4c6fccaa7R132

That change to the type of `filters` fixes lint errors, but causes mismatched-type errors when calling that function in places like this:
https://github.com/matrix-org/matrix-hookshot/blob/main/web/components/roomConfig/GitlabRepoConfig.tsx#L21

```
Argument of type 'GitLabTargetFilter' is not assignable to parameter of type 'Record<string, string> | Record<string, never> | undefined'.
  Type 'GitLabTargetFilter' is not assignable to type 'Record<string, never>'.
```

Basically, I want a type that can cover `{}`, `undefined`, and a `Record<string, string>`, but without using `{}` as a type, which the linter considers an error.

**Comment by AndrewFerr** [comment]
Actually, I'll rebase this before merging, since it was based on the tip of the widgets branch.
